bgpv1: Remove disruptive error handling from BGPRouterManager

YutaroHayakawa opened the pull request:
Currently, we have some disruptive error handling in the `BGPRouterManager`. First in `register()` and the second in `reconcile()` method. Both of them stop the GoBGP instance when it encounters an error in any of the reconcilers in the chain. 

This PR removes the disruptive error-handling from `BGPRouterManager`.

```release-note
bgpv1: Remove disruptive error handling from BGPRouterManager
```

YutaroHayakawa committed:
bgpv1: Don't stop/unregister BGP server on failure

Currently, when the BGP Control Plane meet the error during
reconciliation, it stops and unregisters the server. This comes at a
risk when we accidentally introduce a situation where reconcilers
repeatedly return error for a long time. In such a situation, BGP
Control Plane repeatedly create/destroy server which may cause a BGP
session flapping or frequent route update/withdraw.

We have this problem in two places. The first one is in
BGPRouterManager.register(). When the initial reconciliation fails, BGP
Control Plane stop the server it just creates. In this case, we don't
need to stop server. We can register the server to the RouterManager
before reconciliation and don't have to unregister it even if the
initial reconciliation fails. This is because we have a retry logic now.
The reconciliation failure triggers the retry and in the next cycle, the
reconciliation will be handled in the BGPRouterManager.reconcile()
because the server is already registered.

The second problem is in BGPRouterManager.reconcile(). When the
reconciliation fails, BGP Control Plane stops the BGP server in failure
and unregisters it. Since we have a retry logic, we don't need to stop
BGP servers. We can return error and retry later until we recover from
the failure state.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

harsimran-pabla commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -7,11 +7,13 @@ import (
 	"context"
 	"fmt"
 	"runtime/pprof"
+	"time"
```
I think now we are using cilium/pkg/time instead of std time pkg.

YutaroHayakawa replied:
I removed the retry logic, so this is no longer relevant.

harsimran-pabla commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -187,10 +200,22 @@ func (c *Controller) Run(ctx context.Context) {
 		case <-c.Sig.Sig:
 			l.Info("Cilium BGP Control Plane Controller woken for reconciliation")
+			if retrying {
+				if err := bo.Wait(ctx); err != nil {
```
Should ongoing retry be cancelled if there is new change which triggers reconciliation ?

Main issue I see here is bo.Wait can block for long in between retries. This can cause delay in applying new changes.

YutaroHayakawa replied:
Good point. We currently don't have any way to distinguish the event triggered by retry and the event triggered by configuration change. One way to solve this is putting event origin information into `sig.Event()`. So that we can bypass retry when the event origin is a configuration change.

However, I think it's better, in general, to apply the rate limit here even if the event origin is a configuration change. This is a separate topic, though, because retry and rate limit are different things.

rastislavs replied:
Hmm, this is a though one to decide, but maybe we could apply the same approach as k8s controller-runtime applies, so that we behave similarly to standard k8s controllers. IIRC, in controller-runtime only retries are rate-limited, new changes are reconciled immediately.

See:
 - `c.Queue.AddRateLimited(req)` on error [here](https://github.com/kubernetes-sigs/controller-runtime/blob/b88ed7a3602b85b6cfd0acfe9c25033b978bdb83/pkg/internal/controller/controller.go#L315-L329)
 - `q.Add` (wirthout rate-limiting) on CRUD events [here](https://github.com/kubernetes-sigs/controller-runtime/blob/b88ed7a3602b85b6cfd0acfe9c25033b978bdb83/pkg/handler/enqueue.go#L46-L50)

YutaroHayakawa replied:
As we agreed in the offline conversation, we'll reduce the scope of this PR and discuss separately about retry. Let me close this discussion.

harsimran-pabla commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -187,10 +200,22 @@ func (c *Controller) Run(ctx context.Context) {
 			if err := c.Reconcile(ctx); err != nil {
 				l.WithError(err).Error("Encountered error during reconciliation")
+				c.Sig.Event(struct{}{}) // signal again to retry
+				retrying = true
```
What are the cases where retrying reconciliation will resolve the returned error. We may need to categorize errors which can be retried and which can not. I am not sure if we have practical examples of errors which will be resolved when retried.

YutaroHayakawa replied:
Yeah, this was what I thought during the development, TBH. It is easy to introduce the error classification. We can introduce an error type like [resiliency.Retryable](https://github.com/cilium/cilium/blob/d721b8be7df8372acc3832498638f8c5a699f0f5/pkg/resiliency/error.go#L12), and we can handle it with `errors.Is` and friends.

What we need to decide here is a default behavior (retry by default vs no retry by default). I thought retrying by default was a "safer" option, but yes, practically, we don't have much situation in our current reconcilers that can be recovered by retry. 

The typical case is the error that comes from a temporal IO error (like network or storage), but at the time of reconciliation, we already fetched the k8s resources into the memory, so it doesn't matter; we don't touch the storage; so it doesn't matter. BGP Daemon does retry for any error by itself asynchronously, so it doesn't matter. etc...

In that sense, I don't have any objection to going with no retry by default and introducing `Retryable` error handling later as needed.

rastislavs replied:
I agree on the distinction between retryable and non-retryable errors. And yeah, ATM probably non-retryable could be the default, as majority of our reconcile work is configuring the process-local BGP instance. That would change, if we later decided to go with a BGP process running outside of the agent - I guess in that case most of the errors should be retried.

One error that potentially could be retryable even now is creation of a new gobgp server - that can temporarily fail e.g. if an old instance of the server is still running / was just killed etc. and the bind port is not yet free.

YutaroHayakawa replied:
As we agreed in the offline conversation, we'll reduce the scope of this PR and discuss separately about retry. Let me close this discussion. Currently, we decided to go with no retry by default.

YutaroHayakawa committed:
bgpv1: Run reconcilers twice in the tests

Extend existing reconciler test cases to always execute Reconcile()
twice. This simulates the behavior that BGPRouterManager doesn't clear
the BGP server on reconciliation failure and the next event occurs and
ensures the reconcilers are not relying on that behavior.

As a result, existing reconcilers except NeighborReconciler passed the
test. The fix for the neighbor reconciler will be followed.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bgpv1: Introduce a helper method to construct neighbor ID

There are a lot of repetition of the same fmt.Sprintf in
NeighborReconciler that constructs the same neighbor ID. Introduce a
helper method to construct neighbor ID.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

YutaroHayakawa committed:
bgpv1: Make NeighborReconciler idempotent

Currenly, NeighborReconciler only reconciles on old vs new configuration
differences and doesn't take the actual GoBGP's running state into
account. As a result, when the reconciliation of another reconciler
fails and CurrentServer.Config is not updated, it tries to repeat the
previous reconciliation again. However, when it involves adding the
neighbor, it fails, because adding neighbor is not idempotent in GoBGP.

To solve this issue, we track the neighbor state in metadata and
compare the diff between new config and the state instead of old config.
This makes NeighborReconciler idempotent.

Signed-off-by: Yutaro Hayakawa <yutaro.hayakawa@isovalent.com>

rastislavs reviewed (approved):
LGTM

harsimran-pabla commented on pkg/bgpv1/manager/reconciler/neighbor.go:
```diff
@@ -104,19 +107,19 @@ func (r *NeighborReconciler) Reconcile(ctx context.Context, p ReconcileParams) e
 		if h, ok = nset[key]; !ok {
 			nset[key] = &member{
-				cur: &curNeigh[i],
+				cur: n,
```
Is range issue resolved in latest go versions ? I think in latest go versions `n` is new on each iteration of range. Previously it was not the case so pointer to exact element in the slice was referenced.

YutaroHayakawa replied:
Now `curNeigh` is a pointer slice, so I believe it's safe to copy `n` here.
